.github/workflows/build_and_test_dbt.yaml

```diff
@@ -33,21 +33,35 @@ jobs:
       - name: Configure dbt environment
         uses: ./.github/actions/configure_dbt_environment
 
-      - name: Cache dbt state directory
+      # We have to use the separate `restore`/`save` actions instead of the
+      # unified `cache` action because only `restore` provides access to the
+      # `cache-matched-key` and `cache-primary-key` outputs as of v3
+      - name: Restore dbt state cache
         id: cache
-        uses: actions/cache@v3
+        uses: actions/cache/restore@v3
         with:
           path: ${{ env.PROJECT_DIR }}/${{ env.STATE_DIR }}
           key: ${{ env.CACHE_NAME }}-${{ env.CACHE_KEY }}
           restore-keys: |
             ${{ env.CACHE_NAME }}-master
 
-      - if: ${{ steps.cache.outputs.cache-hit == 'true' }}
+      # If we restore the cache from the `restore-keys` key, the `cache-hit`
+      # output will be 'false' but the `cache-matched-key` output will be
+      # the name of the `restore-keys` key; we want to count this case as a hit
+      - if: |
+          steps.cache.outputs.cache-hit == 'true' ||
+          steps.cache.outputs.cache-matched-key == format(
+            '{0}-master', env.CACHE_NAME
+          )
```

Comment:
[This workflow run](https://github.com/ccao-data/data-architecture/actions/runs/5906337990) has three attempts where I tested each of the possible cases:

1. [Attempt #1](https://github.com/ccao-data/data-architecture/actions/runs/5906337990/attempts/1): Branch cache exists; exact cache key hit ➡️ this block evaluates True
2. [Attempt #2](https://github.com/ccao-data/data-architecture/actions/runs/5906337990/attempts/2): Deleted branch cache, but `master` branch cache exists; partial cache key hit ➡️ this block evaluates True
3. [Attempt #3](https://github.com/ccao-data/data-architecture/actions/runs/5906337990/attempts/3): Deleted both branch cache and `master` branch cache; no cache key hit ➡️ this block evaluates False

Reply:
**praise:** Sheesh, this is very confusing behavior. And weird that the actions maintainers don't provide access to the `cache-` vars in `actions/cache`. Nice work on the fast debugging!